internal/servers/repository.go

```diff
@@ -75,46 +82,114 @@ func (r *Repository) ListServers(ctx context.Context, serverType ServerType, opt
 	return servers, nil
 }
 
+// ServerTag holds the information for the server_tag table for Gorm.
+type ServerTag struct {
+	ServerId string
+	Key      string
+	Value    string
+}
+
+// TableName overrides the table name used by ServerTag to `server_tag`
+func (ServerTag) TableName() string {
+	return "server_tag"
+}
+
+// ListTagsForServers pulls out tag tuples into ServerTag structs for the
+// given server ID values.
+func (r *Repository) ListTagsForServers(ctx context.Context, serverIds []string, opt ...Option) ([]*ServerTag, error) {
+	var serverTags []*ServerTag
+	if err := r.reader.SearchWhere(
+		ctx,
+		&serverTags,
+		"server_id in (?)",
+		[]interface{}{serverIds},
+		db.WithLimit(-1),
+	); err != nil {
+		return nil, fmt.Errorf("error listing server tags with server IDs %#v: %w", serverIds, err)
+	}
+	return serverTags, nil
+}
+
 // UpsertServer adds or updates a server in the DB
 func (r *Repository) UpsertServer(ctx context.Context, server *Server, opt ...Option) ([]*Server, int, error) {
 	if server == nil {
 		return nil, db.NoRowsAffected, errors.New("cannot update server that is nil")
 	}
-	// Ensure, for now at least, the private ID is always equivalent to the name
-	server.PrivateId = server.Name
-	// Build query
-	q := `
-	insert into server
-		(private_id, type, name, description, address, update_time)
-	values
-		($1, $2, $3, $4, $5, $6)
-	on conflict on constraint server_pkey
-	do update set
-		name = $3,
-		description = $4,
-		address = $5,
-		update_time = $6;
-	`
-
-	rowsAffected, err := r.writer.Exec(ctx, q,
-		[]interface{}{
-			server.PrivateId,
-			server.Type,
-			server.Name,
-			server.Description,
-			server.Address,
-			time.Now().Format(time.RFC3339),
-		})
+
+	opts := getOpts(opt...)
+
+	var rowsUpdated int
+	var controllers []*Server
+	_, err := r.writer.DoTx(
+		ctx,
+		db.StdRetryCnt,
+		db.ExpBackoff{},
+		func(read db.Reader, w db.Writer) error {
+			var err error
+			rowsUpdated, err = w.Exec(ctx,
+				serverUpsertQuery,
+				[]interface{}{
+					server.PrivateId,
+					server.Type,
+					server.Description,
+					server.Address,
+					time.Now().Format(time.RFC3339),
+				})
+			if err != nil {
+				return fmt.Errorf("error performing status upsert: %w", err)
+			}
+
+			// If it's a worker, fetch the current controllers to feed to them
+			if server.Type == resource.Worker.String() {
+				// Fetch current controllers to feed to the workers
+				controllers, err = r.listServersWithReader(ctx, read, ServerTypeController)
+				if err != nil {
+					return fmt.Errorf("error fetching list of controllers to send to worker: %w", err)
+				}
+			}
+
+			// If we've been told to update tags, we need to clean out old
+			// ones and add new ones. Within the current transaction, simply
+			// delete all tags for the given worker, then add the new ones
+			// we've been sent.
+			if opts.withUpdateTags {
```

Comment:
It's implicitly tested by anything using tags; if it's not set no tests using tags will work because we'll never insert tags in the first place.

When we get reloading support we'll do before-and-after tests and that will exercise the other half of this.

Reply:
Actually this will get exercised with the tag listing test I'm writing now.